src/arithmetic/arithmetic_expression.c

```diff
@@ -379,80 +381,95 @@ SIValue AR_ADD(SIValue *argv, int argc) {
 }
 
 SIValue AR_SUB(SIValue *argv, int argc) {
-    SIValue result;
-    result = argv[0];
+    SIValue result = argv[0];
+    if (!_validate_numeric(result)) return SI_NullVal();
+
     for(int i = 1; i < argc; i++) {
-        if(SIValue_IsNull(argv[i])) return SI_NullVal();
-        result.doubleval -= argv[i].doubleval;
+        if (!_validate_numeric(argv[i])) return SI_NullVal();
+        result = SIValue_Subtract(result, argv[i]);
     }
     return result;
 }
 
 SIValue AR_MUL(SIValue *argv, int argc) {
-    SIValue result;
-    result = argv[0];
+    SIValue result = argv[0];
+    if (!_validate_numeric(result)) return SI_NullVal();
+
     for(int i = 1; i < argc; i++) {
-        if(SIValue_IsNull(argv[i])) return SI_NullVal();
-        /* Assuming every argv is of type double. */
-        result.doubleval *= argv[i].doubleval;
+        if (!_validate_numeric(argv[i])) return SI_NullVal();
+        result = SIValue_Multiply(result, argv[i]);
     }
     return result;
 }
 
 SIValue AR_DIV(SIValue *argv, int argc) {
-    SIValue result;
-    result = argv[0];
+    SIValue result = argv[0];
+    if (!_validate_numeric(result)) return SI_NullVal();
+
     for(int i = 1; i < argc; i++) {
-        if(SIValue_IsNull(argv[i])) return SI_NullVal();
-        /* Assuming every argv is of type double. */
-        /* TODO: division by zero. */
-        result.doubleval /= argv[i].doubleval;
+        if (!_validate_numeric(argv[i])) return SI_NullVal();
+        result = SIValue_Divide(result, argv[i]);
     }
     return result;
 }
 
+/* TODO All AR_* functions need to emit appropriate failures when provided
+ * with arguments of invalid types and handle multiple arguments. */
+
 SIValue AR_ABS(SIValue *argv, int argc) {
     SIValue result = argv[0];
-    if(result.doubleval < 0) {
-        result.doubleval *= -1;
+    if (!_validate_numeric(result)) return SI_NullVal();
+    switch (SI_TYPE(result)) {
```

Comment:
Thoughts?

if (!_validate_numeric(result)) return SI_NullVal();
if(SI_GET_NUMERIC(argv[0]) < 0) return SIValue_Multiply(argv[0], SI_LongVal(-1));
return argv[0];

Reply:
Yes, I like that better! Good call.